cdap-common/src/main/java/co/cask/cdap/common/security/AuthEnforceRewriter.java

```diff
@@ -0,0 +1,555 @@
+    // AuthEnforce.
+    AuthEnforceAnnotationVisitor classVisitor = new AuthEnforceAnnotationVisitor(className);
+    cr.accept(classVisitor,
+              ClassReader.SKIP_CODE | ClassReader.SKIP_DEBUG | ClassReader.SKIP_FRAMES);
+
+    Map<Method, AnnotationDetails> methodAnnotations = classVisitor.getMethodAnnotations();
+    if (methodAnnotations.isEmpty()) {
+      // if no AuthEnforce annotation was found then return the original class bytes
+      return classBytes;
+    }
+    // We found some method which has AuthEnforce annotation so we need a second pass in to rewrite the class
+    // in second pass we COMPUTE_FRAMES and visit classes with EXPAND_FRAMES
+    ClassWriter cw = new ClassWriter(ClassWriter.COMPUTE_FRAMES);
+    cr.accept(new AuthEnforceAnnotationRewriter(className, cw, classVisitor.getFieldDetails(), methodAnnotations),
+              ClassReader.EXPAND_FRAMES);
+    return cw.toByteArray();
+  }
+
+  /**
+   * A {@link ClassVisitor} which is used in the first pass of the {@link AuthEnforceRewriter} to detect and store
+   * method with {@link AuthEnforce} annotations. Note: This is a visitor, to know the order of in which the below
+   * overridden methods will be called and their overall responsibility please see {@link ClassVisitor} documentation.
+   */
+  private static final class AuthEnforceAnnotationVisitor extends ClassVisitor {
+
+    private final String className;
+    private final Map<Method, AnnotationDetails> methodAnnotations;
+    private boolean interfaceClass;
+    private final Map<String, Type> fieldDetails;
+
+    AuthEnforceAnnotationVisitor(String className) {
+      super(Opcodes.ASM5);
+      this.className = className;
+      this.methodAnnotations = new HashMap<>();
+      this.fieldDetails = new HashMap<>();
+    }
+
+    @Override
+    public void visit(int version, int access, String name, String signature, String superName, String[] interfaces) {
+      // visit the header of the class to identify if this class is an interface
+      super.visit(version, access, name, signature, superName, interfaces);
+      interfaceClass = Modifier.isInterface(access);
+    }
+
+    @Override
+    public FieldVisitor visitField(int access, final String fieldName, String desc, String signature, Object value) {
+      fieldDetails.put(fieldName, Type.getType(desc));
+      return super.visitField(access, fieldName, desc, signature, value);
+    }
+
+    @Override
+    public MethodVisitor visitMethod(final int access, final String methodName, final String methodDesc,
+                                     final String signature, final String[] exceptions) {
+      MethodVisitor mv = super.visitMethod(access, methodName, methodDesc, signature, exceptions);
+
+      // No need to visit the class to look for AuthEnforce annotation if this class is an interface or the method is
+      // static.
+      if (interfaceClass || Modifier.isStatic(access)) {
+        return mv;
```

Comment:
Similar to above, you can `return null` to skip method visitor to speed up. This also means no need to call `super.visitMethod`.